### PR TITLE
Enhance Client Support for Partition Splitting

### PR DESCRIPTION
**Pull request by maochongxin**
## Summary
This merge request introduces enhancements to the client to support partition splitting effectively. The key changes include:
1. Retry Mechanism Enhancement:
+ Improved the retry handling for scenarios where the server undergoes partition splitting. This includes batch operations, LSOP, and both synchronous and asynchronous queries.
2. Proactive Routing Information Refresh:
+ Implemented a feature to refresh routing information based on partition dimensions. This optimization addresses the issue of client blocking caused by refreshing routes in environments with a large number of partitions.

### CODE REVIEW

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryAsyncStreamResult.java [hunk 1]
```diff
@@ -178,10 +224,31 @@ public boolean next() throws Exception {
...
                 Map.Entry<Long, ObPair<Long, ObTableParam>> entry = it.next();
-                // try access new partition, async will not remove useless expectant
-                referToNewPartition(entry.getValue());
+                try {
+                    // try access new partition, async will not remove useless expectant
+                    referToNewPartition(entry.getValue());
+                } catch (Exception e) {
+                    if (e instanceof ObTableNeedFetchAllException) {
+                        this.asyncRequest.getObTableQueryRequest().getTableQuery()
+                            .adjustStartKey(currentStartKey);
```
Comment (shenyunlong) [inline src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryAsyncStreamResult.java:236]: does adjustStartKey work in key-partitioned table?

File: src/main/java/com/alipay/oceanbase/rpc/util/ByteUtil.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+ * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
+ * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+ * See the Mulan PSL v2 for more details.
+ * #L%
+ */
+
+package com.alipay.oceanbase.rpc.util;
+
+public class ByteUtil {
```
Comment (shenyunlong) [inline src/main/java/com/alipay/oceanbase/rpc/util/ByteUtil.java:20]: add ut for ByteUtil

Comment (maochongxin) [reply src/main/java/com/alipay/oceanbase/rpc/util/ByteUtil.java:20]: done

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryAsyncStreamResult.java (hunk 1 above)
Comment (maochongxin) [reply src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryAsyncStreamResult.java]: only range partitioned table is supported; other partitioning uses default behavior

File: src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java [hunk 3]
```diff
@@ -719,6 +774,16 @@ private static TableEntry getTableEntryFromRemote(Connection connection, TableEn
         return tableEntry;
     }
 
+    private static String genLocationSQLByTabletId() {
+        String sql = null;
+        if (ObGlobal.obVsnMajor() >= 4) {
+            sql = PROXY_LOCATION_SQL_PARTITION_BY_TABLETID_V4;
+        } else {
+            throw new FeatureNotSupportedException("not support ob version less than 4");
```
[the remark this answers is not in the document]
Comment (maochongxin) [reply src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java:788]: logic ensures that this `else` is never reached; only the logic for version 4.x will enter here, so this condition might never be triggered

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java [hunk 4]
```diff
@@ -1330,6 +1328,62 @@ private TableEntry refreshTableEntry(TableEntry tableEntry, String tableName)
...
+            if (tableEntry == null) {
+                throw new ObTableEntryRefreshException("Table entry is null, tableName=" + tableName);
+            }
+            long lastRefreshTime = tableEntry.getPartitionEntry().getPartitionInfo(tabletId).getLastUpdateTime();
+            long currentTime = System.currentTimeMillis();
+            if (currentTime - lastRefreshTime < tableEntryRefreshLockTimeout) {
+                return tableEntry;
+            }
+            
+            Lock lock = tableEntry.refreshLockMap.computeIfAbsent(tabletId, k -> new ReentrantLock());
```
[the remark this answers is not in the document]
Comment (maochongxin) [reply src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java:1343]: k -> new ReentrantLock() is a lambda expression that is executed when the key `tabletId` does not exist.

**Pull request merged**